src/libraries/System.Text.Json/src/System/Text/Json/Serialization/WriteStackFrame.cs

```diff
@@ -12,14 +12,14 @@ namespace System.Text.Json
     internal struct WriteStackFrame
     {
         // The object (POCO or IEnumerable) that is being populated.
-        public object CurrentValue;
-        public JsonClassInfo JsonClassInfo;
+        public object? CurrentValue;
+        public JsonClassInfo? JsonClassInfo;
```

Comment:
It is unfortunate that this is marked nullable. All "valid" accesses of WriteStackFrame.JsonClassInfo have `!` after them.

Here's when `JsonClassInfo` can be `null`:
- default instance of WriteStackFrame - this is an internal type, and we never have a default instance
- use of the field after calling `Reset` which is conceptually equivalent to `Dispose` - this should never happen
- Calls to `Push` pass in a `null` `JsonClassInfo`. This should never happen either.

So, as far as I can tell, all code paths/usages of `JsonClassInfo` require using `!`.

**Here are the options:**
1) Leave things as is - nullable field and `!` everywhere - I am fine with this for now until we can refactor this to be better
2) Don't annotate the field as null - this is cleaner but risks potential null ref in the future (though unlikely)
3) Refactor the code (for this specific field) where it becomes more clear/obvious that `JsonClassInfo` cannot be null.

Regarding option 3 (I am only mentioning this here, but am not suggesting we do it here. Let's defer such changes to a subsequent PR):
`JsonClassInfo.ElementClassInfo` can be changed as follows:
```C#
/// <summary>
/// Return the JsonClassInfo for the element type for an enumerable or dictionary.
/// </summary>
/// <remarks>
/// This should not be called during warm-up (initial creation of JsonClassInfos) to avoid recursive behavior
/// which could result in a StackOverflowException.
/// </remarks>
public JsonClassInfo ElementClassInfoOfEnumerableOrDictionary
{
    get
    {
        Debug.Assert(ClassType == ClassType.Enumerable || ClassType == ClassType.Dictionary);

        if (_elementClassInfo == null)
        {
            Debug.Assert(ElementType != null);
            _elementClassInfo = Options.GetOrAddClass(ElementType);
        }

        Debug.Assert(_elementClassInfo != null);
        return _elementClassInfo;
    }
}
```

Similarly, for `JsonPropertyInfo`, we could **add**:
```C#
public JsonClassInfo ElementClassInfoOfEnumerableOrDictionary
{
    get
    {
        Debug.Assert(ClassType == ClassType.Enumerable || ClassType == ClassType.Dictionary);

        if (_elementClassInfo == null)
        {
            Debug.Assert(ElementType != null);
            _elementClassInfo = Options.GetOrAddClass(ElementType);
        }

        Debug.Assert(_elementClassInfo != null);
        return _elementClassInfo;
    }
}
```

Reply:
ok